### PR TITLE
Introduce `do` expressions

### PR DESCRIPTION
**Pull request by hamishknight**
Allow `do` and `do catch` statements to be treated as expressions if they either have single expression branches, or have terminating `then` statements, following the same rules as `if`/`switch` expressions. This is currently gated behind an experimental feature flag pending evolution discussion.

### CODE REVIEW

**Comment by hamishknight** [comment]
https://github.com/apple/swift/pull/68344

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/swift/pull/68344

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/swift/pull/68344

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/swift/pull/68344

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/swift/pull/68344

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/swift/pull/68344

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/swift/pull/68344

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/swift/pull/68344

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/swift/pull/68344

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/swift/pull/68344

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/swift/pull/68344

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/swift/pull/68344

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/swift/pull/68344

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/swift/pull/68344

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/swift/pull/68344

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/swift/pull/68344

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/swift/pull/68344

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/swift/pull/68344

@swift-ci please test Windows

File: Sources/SwiftSyntax/generated/SyntaxBaseNodes.swift [hunk 1]
```diff
@@ -474,6 +474,7 @@ public extension Syntax {
 /// - ``DeclReferenceExprSyntax``
 /// - ``DictionaryExprSyntax``
 /// - ``DiscardAssignmentExprSyntax``
+/// - ``DoExprSyntax``
```
Comment (ahoppen) [inline Sources/SwiftSyntax/generated/SyntaxBaseNodes.swift:477]: Ideally we wouldn’t include experimental nodes here.

File: Sources/SwiftSyntax/generated/SyntaxCollections.swift [hunk 2]
```diff
@@ -229,6 +229,7 @@ public struct AvailabilityArgumentListSyntax: SyntaxCollection, SyntaxHashable {
 ///
 /// ### Contained in
 /// 
+///  - ``DoExprSyntax``.``DoExprSyntax/catchClauses``
```
Comment (ahoppen) [inline Sources/SwiftSyntax/generated/SyntaxCollections.swift:232]: Nor should we include the experimental nodes here.

File: Sources/SwiftParser/Expressions.swift [hunk 3]
```diff
@@ -525,21 +525,26 @@ extension Parser {
...
-    } else if self.at(.keyword(.if)) {
+    case (.if, _)?:
       return RawExprSyntax(
-        parseIfExpression(ifHandle: .constant(.keyword(.if)))
+        self.parseIfExpression(ifHandle: .constant(.keyword(.if)))
       )
+    case (.switch, _)?:
+      return RawExprSyntax(
+        self.parseSwitchExpression(switchHandle: .constant(.keyword(.switch)))
+      )
```
Comment (ahoppen) [inline Sources/SwiftParser/Expressions.swift:545]: I didn’t check but wouldn’t this be cleaner now as

```suggestion
    case (.switch, let handle)?:
      return RawExprSyntax(
        self.parseSwitchExpression(switchHandle: handle)
      )
```

Same for the other `.constant` cases here.

Not that we want to do it, but that would make the code portable to changing `self.at` to `self.canRecoverTo`.

File: Sources/SwiftParser/Expressions.swift [hunk 4]
```diff
@@ -2045,6 +2050,33 @@ extension Parser.Lookahead {
   }
 }
 
+// MARK: Do-Catch Expressions
+
+extension Parser {
+  /// Parse a do expression.
+  mutating func parseDoExpression(doHandle: RecoveryConsumptionHandle) -> RawDoExprSyntax {
+    assert(experimentalFeatures.contains(.doExpressions))
```
Comment (ahoppen) [inline Sources/SwiftParser/Expressions.swift:2058]: I would make this a precondition based on 

https://github.com/apple/swift-syntax/blob/7a643a740411befe13c797ccadbbf372e19bda8e/Sources/SwiftSyntax/Assert.swift#L13-L22

**Review by ahoppen — COMMENTED** [review]
Looks good to me, just a couple of minor comments.

Also: Is there any plan to deprecate/remove `DoStmtSyntax` once `do` expressions have been accepted?

File: Sources/SwiftSyntax/generated/SyntaxBaseNodes.swift (hunk 1 above)
Comment (hamishknight) [reply Sources/SwiftSyntax/generated/SyntaxBaseNodes.swift:477]: Good catch!

File: Sources/SwiftParser/Expressions.swift (hunk 4 above)
Comment (hamishknight) [reply Sources/SwiftParser/Expressions.swift:2058]: Heh, I'm too used to C++ 😅

File: Sources/SwiftParser/Expressions.swift (hunk 3 above)
Comment (hamishknight) [reply Sources/SwiftParser/Expressions.swift:545]: Hah yeah that's what I originally tried, but there didn't seem to be a nice way to go from TokenConsumptionHandle to RecoveryConsumptionHandle, looks like there's `.noRecovery` now, so I should be able to use that.

**Comment by hamishknight** [comment]
> Also: Is there any plan to deprecate/remove DoStmtSyntax once do expressions have been accepted?

Yeah that's the plan

**Comment by hamishknight** [comment]
https://github.com/apple/swift/pull/68344

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/swift/pull/68344

@swift-ci please test

File: Sources/SwiftParser/Expressions.swift (hunk 3 above)
Comment (ahoppen) [reply Sources/SwiftParser/Expressions.swift]: Yeah, I think `noRecovery` is a recent-ish addition.

**Review by ahoppen — APPROVED** [review]

**Pull request merged**